clog/loggers.py

```diff
@@ -295,27 +303,66 @@ def log_line(self, stream, line):
             self.metrics.monk_exception()
 
     def _log_line_no_size_limit(self, stream, line):
-        if time.time() < self.last_timeout + self.timeout_backoff_s:
+        now = time.time()
+
+        if self.buffering:
+            if now - self.buffering_start < self.buffering_time_s:
+                self._add_to_buffer(stream, line)
+                return
+            else:
+                self.buffering = False
+                self._flush_buffer()
+                self.report_status(False, 'Flushed buffer ({} left)'.format(len(self.buffer)))
+
+        if now < self.last_timeout + self.timeout_backoff_s:
             return
+
         with self.metrics.sampled_request():
             try:
                 self.producer.send_messages(
                     self.stream_prefix + stream,
                     [line],
                     None
                 )
-            except socket.timeout as e:
-                self.report_status(True, 'Monk took too long to respond')
-                self.last_timeout = time.time()
-                self.metrics.monk_timeout()
             except Exception as e:
-                self.report_status(
-                    True,
-                    'Exception while sending to monk: %s' % str(e)
-                )
-                self.metrics.monk_exception()
+                if isinstance(e, socket.timeout):
+                    self.report_status(True, 'Monk took too long to respond')
+                    self.metrics.monk_timeout()
+                    self.last_timeout = now
+                else:
+                    self.report_status(True, 'Exception while sending to monk: %s' % str(e))
+                    self.metrics.monk_exception()
+
+                self._start_buffering(now)
+                self._add_to_buffer(stream, line)
+
+    def _start_buffering(self, now):
+        if not self.use_buffer:
+            return
+        self.buffering = True
+        self.report_status(False, 'Start buffering')
+        self.metrics.monk_buffer()
+        self.buffering_start = now
+
+    def _add_to_buffer(self, stream, line):
+        if not self.use_buffer:
+            return
+
+        while self.buffer_bytes + len(line) > self.maximum_buffer_bytes and len(self.buffer) > 0:
+            _, old_line = self.buffer.popleft()
```

Comment:
Is this correct? Does popleft return two items?

Reply:
Nope, but each element is a 2-tuple :)

Reply:
/me facepalm
I'm not sure how i manage to miss that :)